### PR TITLE
SALTO-3741 Add missing features to manifest.xml as optional first

### PR DESCRIPTION
**Pull request by dantal4**
Add missing features as optional first, and if the deploy fails then add it as required.
This way we don't force the target account to have an enable feature, when it is not required.

In the adapter config, features are included as optional by default. In order to include a required feature add the `":required"` suffix to it.
Example:
```
netsuite {
  deploy = {
    additionalDependencies = {
      include = {
        features = [
          "DEPARTMENTS", // will be added as optional
          "SUBSCRIPTIONBILLING:required", // will be added as required
        ]
      }
    }
  }
}
```

---

_Additional context for reviewer_

---
_Release Notes_: 
Netsuite Adapter:
- Add missing features to manifest.xml as optional first
- In the adapter config, features are included as optional by default. In order to include a required feature add the `":required"` suffix to it.

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/netsuite-adapter/src/config.ts [hunk 1]
```diff
@@ -383,27 +393,45 @@ const additionalDependenciesConfigPath: string[] = [
...
 const validateAdditionalDependencies = (
-  { include, exclude }: Record<keyof AdditionalDependencies, unknown>
+  { include, exclude }: Partial<Record<keyof AdditionalDependencies, unknown>>
 ): void => {
+  let validatedInclude: Partial<AdditionalSdfDeployDependencies> = {}
   if (include !== undefined) {
     validatePlainObject(include, additionalDependenciesConfigPath.concat('include'))
     validateAdditionalSdfDeployDependencies(include, 'include')
+    validatedInclude = include
   }
```
Comment (omrilit) [inline packages/netsuite-adapter/src/config.ts:417]: nit: you can extract this code to a typeguard to remove the need of the `validatedInclude`.
```suggestion
  isPartialAdditionalSdfDeployDependencies(include, 'include')
```

File: packages/netsuite-adapter/src/client/client.ts [hunk 2]
```diff
@@ -217,6 +215,63 @@ export default class NetsuiteClient {
...
+      } else if (feature.status === 'optional' && !feature.canBeRequired) {
+        log.warn('The %s feature is required but it is excluded', featureName)
+        missingExcludedFeatures.add(featureName)
+      }
+    })
+    if (missingExcludedFeatures.size > 0) {
+      const error = new Error(`The following features are required but they are excluded: ${Array.from(missingExcludedFeatures).join(', ')}.`)
+      return { failToUpdate: true, error }
+    }
+    return { failToUpdate }
```
Comment (omrilit) [inline packages/netsuite-adapter/src/client/client.ts:272]:
```suggestion
    return { failedToUpdate }
```

File: packages/netsuite-adapter/src/client/client.ts [hunk 3]
```diff
@@ -217,6 +215,63 @@ export default class NetsuiteClient {
...
+  ): FeaturesMap {
+    return Object.fromEntries([
+      ...additionalDependencies.include.features
+        .map(featureName => [featureName, { status: 'optional', canBeRequired: true }]),
+      ...additionalDependencies.exclude.features
+        .map(featureName => [featureName, { status: 'excluded' }]),
+    ])
+  }
+
+  private static updateFeaturesMap(
```
Comment (omrilit) [inline packages/netsuite-adapter/src/client/client.ts:246]: NICE!

File: packages/netsuite-adapter/src/client/client.ts [hunk 4]
```diff
@@ -217,6 +215,63 @@ export default class NetsuiteClient {
...
+    let failToUpdate = false
+    missingFeaturesError.missingFeatures.forEach(featureName => {
+      const feature = featuresMap[featureName]
+      if (feature === undefined) {
+        featuresMap[featureName] = { status: 'optional', canBeRequired: true }
+      } else if (feature.status === 'excluded') {
+        featuresMap[featureName] = { status: 'optional', canBeRequired: false }
+      } else if (feature.status === 'required') {
+        log.error('The %s feature is already required, but sdf returned an error: %o', featureName, missingFeaturesError)
+        failToUpdate = true
```
Comment (omrilit) [inline packages/netsuite-adapter/src/client/client.ts:260]: Why did you decide not to return an error to the caller in this case?
If you'll decide we should return such, then it makes the `failToUpdate` returned value redundant.

File: packages/netsuite-adapter/src/client/manifest_utils.ts [hunk 5]
```diff
@@ -137,34 +137,35 @@ const fixDependenciesObject = (dependencies: Value): void => {
...
   features.feature = _(makeArray(features.feature))
-    // remove all required features
-    .differenceBy(requiredFeatures, item => item[TEXT_ATTRIBUTE])
-    // re-add all required features with "required=true"
-    .unionBy(requiredFeatures, item => item[TEXT_ATTRIBUTE])
-    .filter(item => !additionalDependencies.exclude.features.includes(item[TEXT_ATTRIBUTE]))
+    // remove all additional features
+    .differenceBy(additionalFeatures, getTagText)
+    // re-add all additional features with the desired "required" value for each feature
+    .unionBy(additionalFeatures, getTagText)
```
Comment (omrilit) [inline packages/netsuite-adapter/src/client/manifest_utils.ts:161]: Just to make sure that I follow, is it ok to put only the feature name when it's optional and not using the `{ [REQUIRED_ATTRIBUTE]: 'false', [TEXT_ATTRIBUTE]: feature }` object?
Or I am missing something here?

**Review by omrilit — COMMENTED** [review]
Very nice work!

File: packages/netsuite-adapter/src/client/client.ts (hunk 4 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/client/client.ts:260]: The error that will be returned is the error from SDF, so IMO we shouldn't create another error (not like when a feature was manually excluded, so we want the user to know that).

Comment (omrilit) [reply packages/netsuite-adapter/src/client/client.ts:260]: You can still return the original error from here, no?

File: packages/netsuite-adapter/src/client/manifest_utils.ts (hunk 5 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/client/manifest_utils.ts:161]: You're right-
```
An error occurred during manifest validation.
Details: The required attribute for the SUBSCRIPTIONBILLING feature must be set to true or false in the manifest.
File: ~/manifest.xml
```

File: packages/netsuite-adapter/src/config.ts [hunk 6]
```diff
@@ -404,6 +421,18 @@ const validateAdditionalDependencies = (
     validatePlainObject(exclude, additionalDependenciesConfigPath.concat('exclude'))
     validateAdditionalSdfDeployDependencies(exclude, 'exclude')
   }
+  if (include?.features && exclude?.features) {
+    const conflictedFeatures = _.intersection(include.features, exclude.features)
```
Comment (omrilit) [inline packages/netsuite-adapter/src/config.ts:425]: The `REQUIRED_FEATURE_SUFFIX` should be removed from the included features, no?

**Review by omrilit — APPROVED** [review]
Looks good!
Approving to save cycle.

File: packages/netsuite-adapter/src/config.ts (hunk 6 above)
Comment (dantal4) [reply packages/netsuite-adapter/src/config.ts:425]: right

**Pull request merged**